### ISSUE TITLE
Make it possible to place blocks using copy-and-paste

### ISSUE THREAD

**Issue by cpcallen** [rachel-fenichel/blockly-keyboard-experimentation#103]
It should be possible to:
* navigate to the toolbox (e.g. by pressing `t` or in future `tab`)
* use the cut or copy shortcut to choose a block to place,
* navigate to the desired location on the workspace, and
* use the paste shortcut to paste the chosen block at that location.

**Comment by rachel-fenichel** [comment]
This works in the main workspace but does not work when copying from the flyout, which suggests that the handler for copy needs to be added to the flyout workspace.

The handlers for cut and paste do not apply in the flyout workspace, so they should just be ignored.

**Comment by rachel-fenichel** [comment]
Actually, copy also fires on the flyout but this precondition likely fails: 
```
      preconditionFn: (workspace) => {
        if (
          workspace.keyboardAccessibilityMode &&
          !workspace.options.readOnly
        ) {
          const curNode = workspace.getCursor()?.getCurNode();
          if (curNode && curNode.getSourceBlock()) {
            const sourceBlock = curNode.getSourceBlock();
            return !!(
              !Blockly.Gesture.inProgress() &&
              sourceBlock &&
              sourceBlock.isDeletable() &&
              sourceBlock.isMovable()
            );
          }
        }
        return false;
      },
```

**Comment by rachel-fenichel** [comment]
After https://github.com/google/blockly-keyboard-experimentation/pull/120 it is possible to copy a block from the flyout and paste it to the workspace immediately and without moving focus back to the workspace. It pastes in the most recent cursor location, which was marked when opening the toolbox.

@kmcnaught what do you think the behaviour should be here?

**Comment by kmcnaught** [comment]
I think that pasting in the most recent cursor location is good behaviour - it is clear and not surprising, and if it's not what you wanted then you can presumably navigate the cursor elsewhere before pasting. I think this behaviour should work regardless if you copied from the flyout or the toolbox. Is there another flow I'm missing?

**Comment by kmcnaught** [comment]
I note @cpcallen's comment on #120 about paste from the flyout failing - let me chat with the others here tomorrow. 

Regarding @rachel-fenichel's comment about pasting at the old marker position - I presume this is unintended behaviour that won't be a problem once we lose the concept of marker and cursor being in different places.

**Comment by kmcnaught** [comment]
@rachel-fenichel @cpcallen we discussed this UX flow today. From our perspective (in MakeCode all our blocks are in a flyout, the toolbox just holds categories) we would expect that pressing Ctrl+C on a block in the flyout would:
- report via toast message/notification that you've successfully copied a block
- close the flyout (if it's not fixed)
- focus the ~toolbox~ workspace, with the cursor where you left it (edit 30/01)

This would guide people down the route of "now go where you want to paste it", while avoiding the weirdness of "pasting in the flyout". But I suppose if you navigated back to the flyout yourself for some reason, we'd expect paste to be a no-op there (like pasting into a readonly field)

What do you think?

**Comment by rachel-fenichel** [comment]
I agree with the toast, and ~filed #182 to implement that.~ #124 tracks implementing that.

Closing the flyout if it's auto-closing makes sense.

Paste as a no-op in the flyout makes sense. We may optionally use a toast (same implementation as #182) to tell the user that the paste was a no-op.

@kmcnaught I think instead of focusing the toolbox we should move focus all the way back to the workspace, because there's nowhere to paste a block in the toolbox.

**Comment by kmcnaught** [comment]
@rachel-fenichel I agree, I used the wrong word - glad someone's paying attention

**Comment by kmcnaught** [comment]
> The handlers for cut and paste do not apply in the flyout workspace, so they should just be ignored.

The current behaviour is not quite right here. If you `Ctrl+X` in the flyout, the flyout ignores it but it propagates to the workspace, and cuts the block that has passive focus, which is confusing. I think it should either be a no-op, or treat it as a copy (of the focused flyout block). When the toolbox is focused, the workspace also actions the cut, and it should definitely be a no-op here. 

If you `Ctrl+V` in the toolbox or flyout, it similarly pastes in the workspace. If you were in the flyout, it gets closed and focus is returned to the workspace, so at least the result is clear. If focus is in the toolbox, it may paste behind the flyout (i.e. hard to see) and doesn't move you to the workspace so you might not realise what is happening. I think it would be best for `Ctrl+V` to only work when workspace is focused, but if it does work then the former behaviour makes more sense.

**Comment by rachel-fenichel** [comment]
Changes to focus and selection may have significantly impacted this issue.

Next steps:
Check behaviour for the following cases (ideally by writing tests!) and report which are working and which are broken:

1. Copy **connected** reporter block in main workspace. Paste at:
 a. A block in the workspace with a compatible value input
 b. A block in the workspace with an incompatible value input
 c. A block in the workspace with no value inputs
 d. Anywhere in the flyout
 e. The initial location
 f. Canceled (`Esc`)

1. Copy **connected shadow** reporter block in main workspace. Paste at:
 a. A block in the workspace with a compatible value input
 b. A block in the workspace with an incompatible value input
 c. A block in the workspace with no value inputs
 d. Anywhere in the flyout
 e. The initial location
 f. Canceled (`Esc`)

1. Copy **disconnected** reporter block in main workspace. Paste at:
 a. A block in the workspace with a compatible value input
 b. A block in the workspace with an incompatible value input
 c. A block in the workspace with no value inputs
 d. Anywhere in the flyout
 e. The initial location
 f. Canceled (`Esc`)

1. Copy reporter block in flyout workspace. Paste at:
 a. A block in the workspace with a compatible value input
 b. A block in the workspace with an incompatible value input
 c. A block in the workspace with no value inputs
 d. Anywhere in the flyout
 e. The initial location
 f. Canceled (`Esc`)

1. Copy **connected** stack block in main workspace. Paste at:
 a. A block in the workspace with a compatible next connection
 b. A block in the workspace with a compatible statement input
 c. A block in the workspace with no next connections
 d. Anywhere in the flyout
 e. The initial location
 f. Canceled (`Esc`)

1. Copy **disconnected** stack block in main workspace. Paste at:
 a. A block in the workspace with a compatible next connection
 b. A block in the workspace with a compatible statement input
 c. A block in the workspace with no next connections
 d. Anywhere in the flyout
 e. The initial location
 f. Canceled (`Esc`)

1. Copy stack block in flyout workspace. Paste at:
 a. A block in the workspace with a compatible next connection
 b. A block in the workspace with a compatible statement input
 c. A block in the workspace with no next connections
 d. Anywhere in the flyout
 e. The initial location
 f. Canceled (`Esc`)

Repeat the above for cut/paste flows.

**Comment by maribethb** [comment]
@rachel-fenichel what does "Cancelled" mean in your test cases? If you press "ctrl+c" and a block is successfully copied, at what point do you expect esc to do anything? copy/pasting is not an interruptable process, so I don't think pressing escape should do anything. Am I misunderstanding?

**Comment by rachel-fenichel** [comment]
Canceling is/was only relevant if pasting put a block into move mode immediately. It's irrelevant if the block is just placed without move mode.

**Comment by rachel-fenichel** [comment]
See https://github.com/google/blockly-keyboard-experimentation/issues/480#issuecomment-2905132450 for conversation. 

Remaining issues tracked as individual bugs:
- https://github.com/google/blockly/issues/9091
- https://github.com/google/blockly/issues/9088
- https://github.com/google/blockly-keyboard-experimentation/issues/551
